crash when testing "oklog stream"

This is with oklog 0.2.1. I had run in a seperate window "oklog stream -store log-store-1 -q "JID-" and when I ^C'd I got a crash.

matthew@log-store-1:~$ !okl
oklog store -cluster `hostname --ip-address` -peer log-store-1 -peer log-store-2 -peer log-store-3 -peer log-ingest-1 -peer log-ingest-2 -peer log-ingest-3 --store.path=/data/logs
ts=2017-04-06T13:14:47.361018209Z level=info cluster=10.240.0.24:7659
ts=2017-04-06T13:14:47.361442714Z level=info API=tcp://0.0.0.0:7650
ts=2017-04-06T13:14:47.361859484Z level=info StoreLog=/data/logs
panic: send on closed channel

goroutine 3632 [running]:
github.com/oklog/oklog/pkg/stream.readOnce(0xc9c3c0, 0xc4252b5540, 0xc4252b76a0, 0xc422a77d20, 0x10, 0xc4252b1e60, 0x0, 0x0)
	/Users/peter/src/github.com/oklog/oklog/pkg/stream/stream.go:132 +0x309
github.com/oklog/oklog/pkg/stream.readUntilCanceled(0xc9c3c0, 0xc4252b5540, 0xc4252b76a0, 0xc422a77d20, 0x10, 0xc4252b1e60, 0xa335b0)
	/Users/peter/src/github.com/oklog/oklog/pkg/stream/stream.go:115 +0x61
created by github.com/oklog/oklog/pkg/stream.updateActive
	/Users/peter/src/github.com/oklog/oklog/pkg/stream/stream.go:94 +0x23c

--- a/pkg/store/api.go
+++ b/pkg/store/api.go
@@ -358,23 +358,23 @@ func (a *API) handleUserStream(w http.ResponseWriter, r *http.Request) {
 		return u.String()
 	})
 
-	// The raw chan is closed when the context is canceled.
-	raw := stream.Execute(
-		r.Context(),
-		peerFactory,
-		readCloserFactory,
-		time.Sleep,
-		time.NewTicker,
-	)
+	// Execute returns when the context is canceled.
+	// We must close the raw chan, which we own.
+	raw := make(chan []byte, 1024)
+	go func() {
+		stream.Execute(r.Context(), peerFactory, readCloserFactory, time.Sleep, time.NewTicker, raw)
+		close(raw)
+	}()
 
-	// The deduplicated chan is closed when the raw chan is closed.
-	deduplicated := stream.Deduplicate(
-		raw,
-		window,
-		time.NewTicker,
-	)
+	// Deduplicate returns when the raw chan is closed.
+	// We must close the deduplicated chan, which we own.
+	deduplicated := make(chan []byte, 1024)
+	go func() {
+		stream.Deduplicate(raw, window, time.NewTicker, deduplicated)
+		close(deduplicated)
+	}()
 
-	// Thus, we range over the deduplicated chan.
+	// Thus, we can range over the deduplicated chan.
 	for record := range deduplicated {
 		w.Write(record)
 		w.Write([]byte{'\n'})

--- a/pkg/stream/combined_test.go
+++ b/pkg/stream/combined_test.go
@@ -5,7 +5,6 @@ import (
 	"context"
 	"fmt"
 	"io"
-	"io/ioutil"
 	"math/rand"
 	"testing"
 	"time"
@@ -19,8 +18,8 @@ func TestCombined(t *testing.T) {
 
 	// Our three nodes.
 	type pipe struct {
-		io.Reader
-		io.Writer
+		*io.PipeReader
+		*io.PipeWriter
 	}
 	var (
 		addrs = []string{"foo", "bar", "baz"}
@@ -35,27 +34,31 @@ func TestCombined(t *testing.T) {
 	peerFactory := func() []string {
 		return addrs
 	}
-	readCloserFactory := func(_ context.Context, addr string) (io.ReadCloser, error) {
-		return ioutil.NopCloser(nodes[addr]), nil
+	readCloserFactory := func(ctx context.Context, addr string) (io.ReadCloser, error) {
+		return newContextReadCloser(ctx, struct {
+			io.Reader
+			io.Closer
+		}{
+			Reader: nodes[addr].PipeReader,
+			Closer: nodes[addr].PipeWriter,
+		}), nil
 	}
 
-	// Our stream and deduplicate workers.
+	// Our stream and deduplicate workers, just like in the API.
 	var (
 		ctx, cancel = context.WithCancel(context.Background())
 		window      = 500 * time.Millisecond
 	)
-	raw := Execute(
-		ctx,
-		peerFactory,
-		readCloserFactory,
-		time.Sleep,
-		time.NewTicker,
-	)
-	deduplicated := Deduplicate(
-		raw,
-		window,
-		time.NewTicker,
-	)
+	raw := make(chan []byte, 1024)
+	go func() {
+		Execute(ctx, peerFactory, readCloserFactory, time.Sleep, time.NewTicker, raw)
+		close(raw)
+	}()
+	deduplicated := make(chan []byte, 1024)
+	go func() {
+		Deduplicate(raw, window, time.NewTicker, deduplicated)
+		close(deduplicated)
+	}()
 
 	// We gonna send some unique records.
 	count := 255
@@ -64,12 +67,13 @@ func TestCombined(t *testing.T) {
 	go func() {
 		writers := make([]io.Writer, 0, len(nodes))
 		for _, p := range nodes {
-			writers = append(writers, p.Writer)
+			writers = append(writers, p.PipeWriter)
 		}
 		for i, value := range rand.Perm(count) {
 			id := ulid.MustNew(uint64(value), nil)
 			fmt.Fprintf(writers[(i+0)%len(writers)], "%s %d\n", id, value)
 			fmt.Fprintf(writers[(i+1)%len(writers)], "%s %d\n", id, value)
+			fmt.Fprintf(writers[(i+2)%len(writers)], "%s %d\n", id, value)
 		}
 	}()
 
@@ -91,6 +95,14 @@ func TestCombined(t *testing.T) {
 			t.Errorf("got unexpected record after context cancelation: %q", record)
 		}
 	case <-time.After(time.Second):
-		t.Fatal("timeout waiting for shutdown after context cancelation")
+		t.Fatal("timeout waiting for Execute and Deduplicate shutdown after context cancelation")
 	}
+}
+
+func newContextReadCloser(ctx context.Context, rc io.ReadCloser) io.ReadCloser {
+	go func() {
+		<-ctx.Done()
+		rc.Close()
+	}()
+	return rc
 }

--- a/pkg/stream/deduplicate.go
+++ b/pkg/stream/deduplicate.go
@@ -13,30 +13,25 @@ import (
 // A smaller window may cause duplicate or out-of-order messages.
 // A larger window will cause higher end-to-end latency.
 // The ticker is used every window / 10 to flush the buffer.
-// The returned chan is closed when the in chan is closed.
-func Deduplicate(in <-chan []byte, window time.Duration, ticker func(time.Duration) *time.Ticker) <-chan []byte {
-	out := make(chan []byte, 1024) // TODO(pb): validate buffer size
-	go func() {
-		var (
-			d  = dedupe{BTree: btree.New(2)}
-			tk = ticker(window / 10)
-		)
-		defer tk.Stop()
-		defer close(out)
-		for {
-			select {
-			case record, ok := <-in:
-				if !ok {
-					return
-				}
-				d.insert(record)
-
-			case now := <-tk.C:
-				d.remove(now.Add(-window), out)
+// The function returns when the `in` chan is closed.
+func Deduplicate(in <-chan []byte, window time.Duration, ticker func(time.Duration) *time.Ticker, out chan<- []byte) {
+	var (
+		d  = dedupe{BTree: btree.New(2)}
+		tk = ticker(window / 10)
+	)
+	defer tk.Stop()
+	for {
+		select {
+		case record, ok := <-in:
+			if !ok {
+				return
 			}
+			d.insert(record)
+
+		case now := <-tk.C:
+			d.remove(now.Add(-window), out)
 		}
-	}()
-	return out
+	}
 }
 
 type dedupe struct{ *btree.BTree }

--- a/pkg/stream/deduplicate_test.go
+++ b/pkg/stream/deduplicate_test.go
@@ -17,7 +17,7 @@ func TestDeduplicate(t *testing.T) {
 		tick   = make(chan time.Time)
 		ticker = func(time.Duration) *time.Ticker { return &time.Ticker{C: tick} }
 		window = time.Second
-		out    = Deduplicate(in, window, ticker)
+		out    = make(chan []byte, 1024)
 		t0     = time.Now()
 		t1     = t0.Add(1 * window)
 		t2     = t0.Add(2 * window)
@@ -29,6 +29,12 @@ func TestDeduplicate(t *testing.T) {
 		rec2b = []byte(fmt.Sprintf("%s Bbb2", ulid.MustNew(ulid.Timestamp(t1), nil).String()))
 		rec3  = []byte(fmt.Sprintf("%s Cccc", ulid.MustNew(ulid.Timestamp(t2), nil).String()))
 	)
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		Deduplicate(in, window, ticker, out)
+	}()
 
 	// Out-of-order and duplicate inserts.
 	in <- rec3
@@ -98,9 +104,11 @@ func TestDeduplicate(t *testing.T) {
 		// good
 	}
 
-	// Close the in chan, wait for the close on the out chan.
+	// Close the in chan, wait for Deduplicate to return.
 	close(in)
 	select {
+	case <-done:
+		// good
 	case record, ok := <-out:
 		if ok {
 			t.Fatalf("unexpected record: %q", record)

--- a/pkg/stream/stream.go
+++ b/pkg/stream/stream.go
@@ -20,6 +20,11 @@ type PeerFactory func() []string
 // Other errors will cause the managing goroutine to remanufacture.
 type ReadCloserFactory func(context.Context, string) (io.ReadCloser, error)
 
+type canceldone struct {
+	cancel func()
+	done   <-chan struct{}
+}
+
 // Execute creates and maintains streams of records to multiple peers.
 // It muxes the streams to the returned chan of records.
 // The chan will be closed when the context is canceled.
@@ -37,11 +42,11 @@ func Execute(
 ) <-chan []byte {
 	// Make the sink chan of records.
 	// TODO(pb): validate the buffer size
-	c := make(chan []byte, 1024)
+	sink := make(chan []byte, 1024)
 
 	// Invoke the PeerFactory to get the initial addrs.
 	// Initialize connection managers to each of them.
-	active := updateActive(ctx, nil, pf(), rcf, c, sleep)
+	active := updateActive(ctx, nil, pf(), rcf, sink, sleep)
 
 	go func() {
 		// Re-invoke the peerFactory every second.
@@ -54,52 +59,62 @@ func Execute(
 			case <-tk.C:
 				// Detect new peers, and create connection managers for them.
 				// Terminate connection managers for peers that have gone away.
-				active = updateActive(ctx, active, pf(), rcf, c, sleep)
+				active = updateActive(ctx, active, pf(), rcf, sink, sleep)
 
 			case <-ctx.Done():
 				// Context cancelation is transitive.
-				// We just need to exit.
-				close(c)
+				// We just need to wait.
+				for _, cd := range active {
+					<-cd.done
+				}
+				close(sink)
 				return
 			}
 		}
 	}()
 
-	return c
+	return sink
 }
 
 func updateActive(
 	parent context.Context,
-	prevgen map[string]func(),
+	prevgen map[string]canceldone,
 	addrs []string,
 	rcf ReadCloserFactory,
 	sink chan<- []byte,
 	sleep func(time.Duration),
-) map[string]func() {
+) map[string]canceldone {
 	// Create the "new" collection of peer managers.
 	// Really, we just have to track the cancel func.
-	nextgen := map[string]func(){}
+	nextgen := map[string]canceldone{}
 
 	// The addrs represent all the connections we *should* have.
 	for _, addr := range addrs {
-		if cancel, ok := prevgen[addr]; ok {
+		if cd, ok := prevgen[addr]; ok {
 			// This addr already exists in our previous collection.
 			// Just move its cancel func over to the new collection.
-			nextgen[addr] = cancel
+			nextgen[addr] = cd
 			delete(prevgen, addr)
 		} else {
 			// This addr appears to be new!
 			// Create a new connection manager for it.
-			ctx, cancel := context.WithCancel(parent)
-			go readUntilCanceled(ctx, rcf, addr, sink, sleep)
-			nextgen[addr] = cancel
+			var (
+				ctx, cancel = context.WithCancel(parent)
+				done        = make(chan struct{})
+			)
+			go func(addr string) {
+				readUntilCanceled(ctx, rcf, addr, sink, sleep)
+				close(done)
+			}(addr)
+			nextgen[addr] = canceldone{cancel, done}
 		}
 	}
 
 	// All the addrs left over in the previous collection are gone.
 	// Their connection managers should be canceled.
-	for _, cancel := range prevgen {
-		cancel()
+	for _, cd := range prevgen {
+		cd.cancel()
+		<-cd.done
 	}
 
 	// Good to go.

--- a/pkg/stream/stream.go
+++ b/pkg/stream/stream.go
@@ -26,11 +26,9 @@ type canceldone struct {
 }
 
 // Execute creates and maintains streams of records to multiple peers.
-// It muxes the streams to the returned chan of records.
-// The chan will be closed when the context is canceled.
+// It muxes the streams of incoming records to the sink chan of records.
 // It's designed to be invoked once per user stream request.
 //
-// Incoming records are muxed onto the provided sink chan.
 // The sleep func is used to backoff between retries of a single peer.
 // The ticker func is used to regularly resolve peers.
 func Execute(
@@ -39,41 +37,33 @@ func Execute(
 	rcf ReadCloserFactory,
 	sleep func(time.Duration),
 	ticker func(time.Duration) *time.Ticker,
-) <-chan []byte {
-	// Make the sink chan of records.
-	// TODO(pb): validate the buffer size
-	sink := make(chan []byte, 1024)
-
+	sink chan<- []byte,
+) error {
 	// Invoke the PeerFactory to get the initial addrs.
 	// Initialize connection managers to each of them.
 	active := updateActive(ctx, nil, pf(), rcf, sink, sleep)
 
-	go func() {
-		// Re-invoke the peerFactory every second.
-		// This catches changes in topology.
-		tk := ticker(time.Second)
-		defer tk.Stop()
+	// Re-invoke the peerFactory every second.
+	// This catches changes in topology.
+	tk := ticker(time.Second)
+	defer tk.Stop()
 
-		for {
-			select {
-			case <-tk.C:
-				// Detect new peers, and create connection managers for them.
-				// Terminate connection managers for peers that have gone away.
-				active = updateActive(ctx, active, pf(), rcf, sink, sleep)
+	for {
+		select {
+		case <-tk.C:
+			// Detect new peers, and create connection managers for them.
+			// Terminate connection managers for peers that have gone away.
+			active = updateActive(ctx, active, pf(), rcf, sink, sleep)
 
-			case <-ctx.Done():
-				// Context cancelation is transitive.
-				// We just need to wait.
-				for _, cd := range active {
-					<-cd.done
-				}
-				close(sink)
-				return
+		case <-ctx.Done():
+			// Context cancelation is transitive.
+			// We just need to wait.
+			for _, cd := range active {
+				<-cd.done
 			}
+			return ctx.Err()
 		}
-	}()
-
-	return sink
+	}
 }
 
 func updateActive(
@@ -127,15 +117,18 @@ func updateActive(
 // readUntilCanceled blocks until the context is canceled.
 func readUntilCanceled(ctx context.Context, rcf ReadCloserFactory, addr string, sink chan<- []byte, sleep func(time.Duration)) {
 	for {
-		switch readOnce(ctx, rcf, addr, sink) {
-		case context.Canceled:
-			return // fatal
+		readOnce(ctx, rcf, addr, sink)
+		select {
+		case <-ctx.Done():
+			return
 		default:
 			sleep(time.Second) // TODO(pb): better strategy?
 		}
 	}
 }
 
+// readOnce uses rcf to construct a ReadCloser to the given addr, and consumes
+// and forwards records to the sink until
 func readOnce(ctx context.Context, rcf ReadCloserFactory, addr string, sink chan<- []byte) error {
 	rc, err := rcf(ctx, addr)
 	if err != nil {

--- a/pkg/stream/stream_test.go
+++ b/pkg/stream/stream_test.go
@@ -87,6 +87,49 @@ func TestReadUntilCanceled(t *testing.T) {
 	}
 }
 
+func TestIssue59(t *testing.T) {
+	// Make two sources.
+	rcf := func(_ context.Context, peer string) (io.ReadCloser, error) {
+		switch peer {
+		case "a":
+			return infiniteReader("alpha"), nil
+		case "b":
+			return infiniteReader("beta"), nil
+		default:
+			panic("invalid peer")
+		}
+	}
+
+	// Start Execute, streaming from the two sources.
+	var (
+		ctx, cancel = context.WithCancel(context.Background())
+		pf          = func() []string { return []string{"a", "b"} }
+		sleep       = func(time.Duration) {}
+		tickc       = make(chan time.Time)
+		ticker      = func(time.Duration) *time.Ticker { return &time.Ticker{C: tickc} }
+		sink        = Execute(ctx, pf, rcf, sleep, ticker)
+	)
+
+	// Drain the records as quickly as possible.
+	var (
+		draincomplete = make(chan struct{})
+		drainrecords  = 0
+	)
+	go func() {
+		defer close(draincomplete)
+		for range sink {
+			drainrecords++
+		}
+	}()
+
+	// Run it awhile, then cancel.
+	begin := time.Now()
+	time.Sleep(100 * time.Millisecond)
+	cancel()
+	<-draincomplete
+	t.Logf("drained %d in %s", drainrecords, time.Since(begin))
+}
+
 type ctxReader struct {
 	ctx context.Context
 	rec []byte
@@ -106,3 +149,11 @@ func (r *ctxReader) Read(p []byte) (int, error) {
 }
 
 func (*ctxReader) Close() error { return nil }
+
+type infiniteReader string
+
+func (r infiniteReader) Read(p []byte) (int, error) {
+	return copy(p, []byte(string(r)+"\n")), nil
+}
+
+func (r infiniteReader) Close() error { return nil }

--- a/pkg/stream/stream_test.go
+++ b/pkg/stream/stream_test.go
@@ -95,16 +95,23 @@ func TestIssue59(t *testing.T) {
 		rcf         = func(_ context.Context, peer string) (io.ReadCloser, error) { return infiniteReader(peer), nil }
 		sleep       = func(time.Duration) {} // no sleep
 		ticker      = time.NewTicker
-		sink        = Execute(ctx, pf, rcf, sleep, ticker)
+		sink        = make(chan []byte, 1024)
+		done        = make(chan struct{})
 	)
+
+	go func() {
+		Execute(ctx, pf, rcf, sleep, ticker, sink)
+		close(sink)
+		close(done)
+	}()
 
 	// Drain the records as quickly as possible.
 	var (
-		done  = make(chan struct{})
-		count = 0
+		draindone = make(chan struct{})
+		count     = 0
 	)
 	go func() {
-		defer close(done)
+		defer close(draindone)
 		for range sink {
 			count++
 		}
@@ -112,10 +119,11 @@ func TestIssue59(t *testing.T) {
 
 	// Run it awhile, then cancel.
 	begin := time.Now()
-	time.Sleep(100 * time.Millisecond)
+	time.Sleep(250 * time.Millisecond)
 	cancel()
-	<-done
+	<-draindone
 	t.Logf("drained %d in %s", count, time.Since(begin))
+	<-done
 }
 
 type ctxReader struct {

--- a/pkg/stream/stream_test.go
+++ b/pkg/stream/stream_test.go
@@ -88,37 +88,25 @@ func TestReadUntilCanceled(t *testing.T) {
 }
 
 func TestIssue59(t *testing.T) {
-	// Make two sources.
-	rcf := func(_ context.Context, peer string) (io.ReadCloser, error) {
-		switch peer {
-		case "a":
-			return infiniteReader("alpha"), nil
-		case "b":
-			return infiniteReader("beta"), nil
-		default:
-			panic("invalid peer")
-		}
-	}
-
-	// Start Execute, streaming from the two sources.
+	// Start Execute, streaming from some infinite sources.
 	var (
 		ctx, cancel = context.WithCancel(context.Background())
-		pf          = func() []string { return []string{"a", "b"} }
-		sleep       = func(time.Duration) {}
-		tickc       = make(chan time.Time)
-		ticker      = func(time.Duration) *time.Ticker { return &time.Ticker{C: tickc} }
+		pf          = func() []string { return []string{"alpha", "bravo", "charlie"} }
+		rcf         = func(_ context.Context, peer string) (io.ReadCloser, error) { return infiniteReader(peer), nil }
+		sleep       = func(time.Duration) {} // no sleep
+		ticker      = time.NewTicker
 		sink        = Execute(ctx, pf, rcf, sleep, ticker)
 	)
 
 	// Drain the records as quickly as possible.
 	var (
-		draincomplete = make(chan struct{})
-		drainrecords  = 0
+		done  = make(chan struct{})
+		count = 0
 	)
 	go func() {
-		defer close(draincomplete)
+		defer close(done)
 		for range sink {
-			drainrecords++
+			count++
 		}
 	}()
 
@@ -126,8 +114,8 @@ func TestIssue59(t *testing.T) {
 	begin := time.Now()
 	time.Sleep(100 * time.Millisecond)
 	cancel()
-	<-draincomplete
-	t.Logf("drained %d in %s", drainrecords, time.Since(begin))
+	<-done
+	t.Logf("drained %d in %s", count, time.Since(begin))
 }
 
 type ctxReader struct {

